### PR TITLE
The bufferstate vector normally has 8k elements. Since the BufferStat…

### PR DESCRIPTION
**Pull request by baldersheim**
…e was 160 bytes this is aroung 1.2M.

This reduces the BufferState to 128 making it fit nicely in a 1M allocation.
However the cost is that you get an extra small allocation for the _freeList. We should see if we could reclain that by further reducing the size of the BufferState.
Perhaps the element count will be sufficient with uint32_t.

@toregge and @geirst PR

### CODE REVIEW

File: searchlib/src/vespa/searchlib/attribute/enum_store_dictionary.cpp [hunk 1]
```diff
@@ -310,3 +310,11 @@ class BTree<IEnumStore::Index, uint32_t, NoAggregated,
...
 
 }
+
+namespace vespalib::datastore {
+
+using namespace btree;
+
+VESPALIB_DATASTORE_INSTANTIATE_BUFFERTYPE_INTERNALNODE(EntryRef, NoAggregated, BTreeDefaultTraits::INTERNAL_SLOTS);
+
+}
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/attribute/enum_store_dictionary.cpp:320]: Missing newline at end of file.

File: storage/src/vespa/storage/bucketdb/btree_lockable_map.cpp [hunk 2]
```diff
@@ -1,8 +1,15 @@
...
 
 template class BTreeLockableMap<StorageBucketInfo>; // Forced instantiation.
 
 }
+
+namespace vespalib::datastore {
+
+template class BufferType<storage::bucketdb::StorageBucketInfo>;
+
+}
```
Comment (toregge) [inline storage/src/vespa/storage/bucketdb/btree_lockable_map.cpp:15]: Missing newline at end of file.

File: vespalib/src/vespa/vespalib/btree/btreenodestore.cpp [hunk 3]
```diff
@@ -18,4 +19,33 @@ template class BTreeNodeStore<uint32_t, int32_t, MinMaxAggregated,
...
+VESPALIB_DATASTORE_INSTANTIATE_BUFFERTYPE_LEAFNODE(EntryRef, uint32_t, NoAggregated, BTreeDefaultTraits::LEAF_SLOTS);
+VESPALIB_DATASTORE_INSTANTIATE_BUFFERTYPE_LEAFNODE(EntryRef, BTreeNoLeafData, NoAggregated, BTreeDefaultTraits::LEAF_SLOTS);
+VESPALIB_DATASTORE_INSTANTIATE_BUFFERTYPE_LEAFNODE(EntryRef, EntryRef, NoAggregated, BTreeDefaultTraits::LEAF_SLOTS);
+
+template class BufferType<BTreeKeyData<uint32_t, uint32_t>>;
+template class BufferType<BTreeKeyData<uint32_t, int32_t>>;
+template class BufferType<BTreeKeyData<uint32_t, BTreeNoLeafData>>;
+template class BufferType<BTreeKeyData<uint32_t, EntryRef>>;
+
+}
```
Comment (toregge) [inline vespalib/src/vespa/vespalib/btree/btreenodestore.cpp:51]: Missing newline at end of file.

File: vespalib/src/vespa/vespalib/btree/btreenodestore.h [hunk 4]
```diff
@@ -219,4 +222,33 @@ extern template class BTreeNodeStore<uint32_t, int32_t, MinMaxAggregated,
...
+extern template class BufferType<BTreeLeafNode<uint32_t, BTreeNoLeafData, NoAggregated, BTreeDefaultTraits::LEAF_SLOTS>,
+                                 FrozenBtreeNode<BTreeLeafNode<uint32_t, BTreeNoLeafData, NoAggregated, BTreeDefaultTraits::LEAF_SLOTS>>>;
+extern template class BufferType<BTreeLeafNode<uint32_t, int32_t, MinMaxAggregated, BTreeDefaultTraits::LEAF_SLOTS>,
+                                 FrozenBtreeNode<BTreeLeafNode<uint32_t, int32_t, MinMaxAggregated, BTreeDefaultTraits::LEAF_SLOTS>>>;
+
+extern template class BufferType<BTreeKeyData<uint32_t, uint32_t>>;
+extern template class BufferType<BTreeKeyData<uint32_t, int32_t>>;
+extern template class BufferType<BTreeKeyData<uint32_t, BTreeNoLeafData>>;
+
+}
```
Comment (toregge) [inline vespalib/src/vespa/vespalib/btree/btreenodestore.h:254]: Missing newline at end of file.

File: vespalib/src/vespa/vespalib/btree/btreestore.cpp [hunk 5]
```diff
@@ -11,3 +11,13 @@ template class BTreeStore<uint32_t, BTreeNoLeafData, NoAggregated, std::less<uin
...
+
+namespace vespalib::datastore {
+
+using namespace btree;
+
+template class BufferType<BTreeRoot<uint32_t, uint32_t, NoAggregated, std::less<uint32_t>, BTreeDefaultTraits>>;
+template class BufferType<BTreeRoot<uint32_t, BTreeNoLeafData, NoAggregated, std::less<uint32_t>, BTreeDefaultTraits>>;
+template class BufferType<BTreeRoot<uint32_t, int32_t, MinMaxAggregated, std::less<uint32_t>, BTreeDefaultTraits, MinMaxAggrCalc>>;
+
+}
```
Comment (toregge) [inline vespalib/src/vespa/vespalib/btree/btreestore.cpp:23]: Missing newline at end of file.

File: vespalib/src/vespa/vespalib/datastore/buffer_type.h [hunk 6]
```diff
@@ -2,11 +2,12 @@
...
 #pragma once
 
-#include <cstdint>
-#include <cstddef>
+#include "atomic_entry_ref.h"
+#include <string>
 
 namespace vespalib::datastore {
 
+using ElemCount = uint32_t;
```
Comment (toregge) [inline vespalib/src/vespa/vespalib/datastore/buffer_type.h:10]: Wrong type for ElemCount. A buffer can contain more than 4G elements. E.g. when using 1024 buffers and array size > 1024. Clamping `_maxArrays`, `_minArrays` and `_numArraysForNewBuffer` in `BufferTypeBase` might work.

**Review by toregge — CHANGES_REQUESTED** [review]
Most data stores have 1024 buffers.  ArrayStore is the exception (with 8192 buffers due to array size for small arrays being a property of the the buffer type).

File: vespalib/src/vespa/vespalib/datastore/array_store.cpp [hunk 7]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/array_store.cpp:1]: Please use new copyright and year 2021.

File: vespalib/src/vespa/vespalib/datastore/buffer_type.hpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/buffer_type.hpp:1]: Please use new copyright and year 2021.

File: vespalib/src/vespa/vespalib/datastore/unique_store.cpp [hunk 9]
```diff
@@ -0,0 +1,21 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/unique_store.cpp:1]: Please use new copyright and year 2021.

File: vespalib/src/vespa/vespalib/datastore/array_store.cpp (hunk 7 above)
Comment (kkraune) [reply vespalib/src/vespa/vespalib/datastore/array_store.cpp:1]: year is not mandated, makes it easier for later

File: searchlib/src/vespa/searchlib/attribute/enum_store_dictionary.cpp (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/enum_store_dictionary.cpp]: Fixed

Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/enum_store_dictionary.cpp]: Fixed

File: storage/src/vespa/storage/bucketdb/btree_lockable_map.cpp (hunk 2 above)
Comment (baldersheim) [reply storage/src/vespa/storage/bucketdb/btree_lockable_map.cpp]: Fixed

File: vespalib/src/vespa/vespalib/btree/btreenodestore.cpp (hunk 3 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/btree/btreenodestore.cpp]: Fixed

File: vespalib/src/vespa/vespalib/btree/btreenodestore.h (hunk 4 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/btree/btreenodestore.h]: Fixed

File: vespalib/src/vespa/vespalib/btree/btreestore.cpp (hunk 5 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/btree/btreestore.cpp]: Fixed

File: vespalib/src/vespa/vespalib/datastore/array_store.cpp (hunk 7 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/datastore/array_store.cpp]: Fixed

File: vespalib/src/vespa/vespalib/datastore/unique_store.cpp (hunk 9 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/datastore/unique_store.cpp]: Fixed

File: vespalib/src/vespa/vespalib/datastore/buffer_type.hpp (hunk 8 above)
Comment (baldersheim) [reply vespalib/src/vespa/vespalib/datastore/buffer_type.hpp]: Fixed

**Comment by baldersheim** [comment]
Superceeded by #15980

**Pull request closed without merging**